Warn for annotated spectra with unknown sequence formatting

Anyone got a guess as to what this error might mean?  I am training on an MGF containing 2,373,619 spectra with Casanovo v3.4.0. According to the stderr output, the training read in 2,373,619 spectra and then died with this traceback:

```
Traceback (most recent call last):
  File "/net/noble/vol1/home/noble/miniconda3/envs/casanovo/bin/casanovo", line 8, in <module>
    sys.exit(main())
  File "/net/noble/vol1/home/noble/miniconda3/envs/casanovo/lib/python3.10/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/net/noble/vol1/home/noble/miniconda3/envs/casanovo/lib/python3.10/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/net/noble/vol1/home/noble/miniconda3/envs/casanovo/lib/python3.10/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/net/noble/vol1/home/noble/miniconda3/envs/casanovo/lib/python3.10/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/net/noble/vol1/home/noble/miniconda3/envs/casanovo/lib/python3.10/site-packages/casanovo/casanovo.py", line 173, in main
    model_runner.train(peak_path, peak_path_val, model, config)
  File "/net/noble/vol1/home/noble/miniconda3/envs/casanovo/lib/python3.10/site-packages/casanovo/denovo/model_runner.py", line 229, in train
    train_index = AnnotatedSpectrumIndex(
  File "/net/noble/vol1/home/noble/miniconda3/envs/casanovo/lib/python3.10/site-packages/depthcharge/data/hdf5.py", line 445, in __init__
    super().__init__(
  File "/net/noble/vol1/home/noble/miniconda3/envs/casanovo/lib/python3.10/site-packages/depthcharge/data/hdf5.py", line 104, in __init__
    self.add_file(ms_file)
  File "/net/noble/vol1/home/noble/miniconda3/envs/casanovo/lib/python3.10/site-packages/depthcharge/data/hdf5.py", line 230, in add_file
    group.create_dataset(
  File "/net/noble/vol1/home/noble/miniconda3/envs/casanovo/lib/python3.10/site-packages/h5py/_hl/group.py", line 183, in create_dataset
    dsid = dataset.make_new_dset(group, shape, dtype, data, name, **kwds)
  File "/net/noble/vol1/home/noble/miniconda3/envs/casanovo/lib/python3.10/site-packages/h5py/_hl/dataset.py", line 168, in make_new_dset
    dset_id.write(h5s.ALL, h5s.ALL, data)
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "h5py/h5d.pyx", line 280, in h5py.h5d.DatasetID.write
  File "h5py/_proxy.pyx", line 145, in h5py._proxy.dset_rw
  File "h5py/_conv.pyx", line 444, in h5py._conv.str2vlen
  File "h5py/_conv.pyx", line 95, in h5py._conv.generic_converter
  File "h5py/_conv.pyx", line 249, in h5py._conv.conv_str2vlen
TypeError: Can't implicitly convert non-string objects to strings
```

FWIW, the log file is attached.
[decoy.log](https://github.com/Noble-Lab/casanovo/files/12004518/decoy.log)

It looks like this same (or a very similar) error is thrown when the ` AnnotatedSpectrumIndex` constructor is called with an unannotated MGF file in `ms_data_files`. Here's the output from that specific case:

```
casanovo\denovo\model_runner.py:183: in predict
    test_index = self._get_index(peak_path, evaluate, "")
casanovo\denovo\model_runner.py:420: in _get_index
    return Index(index_fname, filenames, valid_charge=valid_charge)
D:\anaconda3\envs\casanovo_env\lib\site-packages\depthcharge\data\hdf5.py:446: in __init__
    super().__init__(
D:\anaconda3\envs\casanovo_env\lib\site-packages\depthcharge\data\hdf5.py:104: in __init__
    self.add_file(ms_file)
D:\anaconda3\envs\casanovo_env\lib\site-packages\depthcharge\data\hdf5.py:231: in add_file
    group.create_dataset(
D:\anaconda3\envs\casanovo_env\lib\site-packages\h5py\_hl\group.py:183: in create_dataset
    dsid = dataset.make_new_dset(group, shape, dtype, data, name, **kwds)
D:\anaconda3\envs\casanovo_env\lib\site-packages\h5py\_hl\dataset.py:166: in make_new_dset
    dset_id.write(h5s.ALL, h5s.ALL, data)
h5py\\_objects.pyx:54: in h5py._objects.with_phil.wrapper
    ???
h5py\\_objects.pyx:55: in h5py._objects.with_phil.wrapper
    ???
h5py\\h5d.pyx:282: in h5py.h5d.DatasetID.write
    ???
h5py\\_proxy.pyx:147: in h5py._proxy.dset_rw
    ???
h5py\\_conv.pyx:442: in h5py._conv.str2vlen
    ???
h5py\\_conv.pyx:96: in h5py._conv.generic_converter
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   ???
E   TypeError: Can't implicitly convert non-string objects to strings

h5py\\_conv.pyx:247: TypeError
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -14,6 +14,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 
 - Increased minimum Python version from 3.8 to 3.10.
 - DepthCharge is upgraded to the latest version 0.4.9.
+- A more descriptive error message will now be logged for some annotated spectrum file parsing failure cases.
 
 ### Fixed
 

--- a/casanovo/denovo/model_runner.py
+++ b/casanovo/denovo/model_runner.py
@@ -294,12 +294,13 @@ class ModelRunner:
 
         try:
             self.loaders.setup(stage="test", annotated=evaluate)
-        except (KeyError, OSError) as e:
+        except (KeyError, OSError, TypeError) as e:
             if evaluate:
                 error_message = (
                     "Error creating annotated spectrum dataloaders. This may "
                     "be the result of having an unannotated peak file present "
-                    "in the validation peak file path list."
+                    "in the validation peak file path list. Check that the "
+                    "spectrum files are annotated correctly."
                 )
 
                 logger.error(error_message)

--- a/casanovo/denovo/model_runner.py
+++ b/casanovo/denovo/model_runner.py
@@ -299,7 +299,7 @@ class ModelRunner:
                 error_message = (
                     "Error creating annotated spectrum dataloaders. This may "
                     "be the result of having an unannotated peak file present "
-                    "in the validation peak file path list. Check that the input"
+                    "in the validation peak file path list. Check that the input "
                     "spectrum files are annotated, and if they are that the "
                     "format is correct."
                 )

--- a/casanovo/denovo/model_runner.py
+++ b/casanovo/denovo/model_runner.py
@@ -299,8 +299,9 @@ class ModelRunner:
                 error_message = (
                     "Error creating annotated spectrum dataloaders. This may "
                     "be the result of having an unannotated peak file present "
-                    "in the validation peak file path list. Check that the "
-                    "spectrum files are annotated correctly."
+                    "in the validation peak file path list. Check that the input"
+                    "spectrum files are annotated, and if they are that the "
+                    "format is correct."
                 )
 
                 logger.error(error_message)

--- a/tests/unit_tests/test_runner.py
+++ b/tests/unit_tests/test_runner.py
@@ -256,92 +256,120 @@ def test_save_final_model(tmp_path, mgf_small, tiny_config):
     assert validation_file.exists()
 
 
-def test_evaluate(
-    tmp_path, mgf_small, mzml_small, mgf_small_unannotated, tiny_config
-):
-    """Test model evaluation during sequencing"""
-    # Train tiny model
+def test_evaluate_success(tmp_path, mgf_small, tiny_config):
+    """Test successful model evaluation with annotated peak files."""
     config = Config(tiny_config)
     config.max_epochs = 1
+
     model_file = tmp_path / "epoch=0-step=1.ckpt"
+    result_file = tmp_path / "result.mztab"
+
     with ModelRunner(config, output_dir=tmp_path) as runner:
         runner.train([mgf_small], [mgf_small])
 
     assert model_file.is_file()
 
-    # Test evaluation with annotated peak file
-    result_file = tmp_path / "result.mztab"
     with ModelRunner(
         config, model_filename=str(model_file), overwrite_ckpt_check=False
     ) as runner:
         runner.predict([mgf_small], result_file, evaluate=True)
 
     assert result_file.is_file()
-    result_file.unlink()
 
+
+@pytest.mark.parametrize(
+    "input_files, expect_match",
+    [
+        pytest.param(
+            ["mzml_small"],
+            False,
+            id="mzml_only",
+        ),
+        pytest.param(
+            ["mgf_small_unannotated"],
+            True,
+            id="unannotated_mgf_only",
+        ),
+        pytest.param(
+            ["mgf_small_unannotated", "mzml_small"],
+            True,
+            id="unannotated_mgf_and_mzml",
+        ),
+        pytest.param(
+            ["mgf_small", "mzml_small"],
+            False,
+            id="annotated_mgf_and_mzml",
+        ),
+        pytest.param(
+            ["mgf_small", "mgf_small_unannotated"],
+            True,
+            id="annotated_and_unannotated_mgf",
+        ),
+        pytest.param(
+            ["mgf_small", "mgf_small_unannotated", "mzml_small"],
+            True,
+            id="annotated_unannotated_and_mzml",
+        ),
+        pytest.param(
+            ["improperly_annotated_file"],
+            True,
+            id="improperly_annotated_mgf_missing_seq",
+        ),
+    ],
+)
+def test_evaluate_failure_cases(
+    tmp_path,
+    mgf_small,
+    mzml_small,
+    mgf_small_unannotated,
+    tiny_config,
+    input_files,
+    expect_match,
+):
+    """Test all evaluation failure modes."""
     exception_string = (
         "Error creating annotated spectrum dataloaders. This may "
         "be the result of having an unannotated peak file present "
         "in the validation peak file path list."
     )
 
-    with pytest.raises(TypeError):
+    config = Config(tiny_config)
+    config.max_epochs = 1
+
+    model_file = tmp_path / "epoch=0-step=1.ckpt"
+    result_file = tmp_path / "result.mztab"
+
+    with ModelRunner(config, output_dir=tmp_path) as runner:
+        runner.train([mgf_small], [mgf_small])
+
+    assert model_file.is_file()
+
+    mgf_path = Path(mgf_small)
+    improperly_annotated_file = mgf_path.parent / "improperly_annotated.mgf"
+    improperly_annotated_file.write_text(
+        mgf_path.read_text().replace("SEQ=", "PEPTIDE=")
+    )
+
+    file_map = {
+        "mgf_small": mgf_small,
+        "mzml_small": mzml_small,
+        "mgf_small_unannotated": mgf_small_unannotated,
+        "improperly_annotated_file": improperly_annotated_file,
+    }
+
+    files = [file_map[name] for name in input_files]
+
+    with pytest.raises(
+        TypeError,
+        match=exception_string if expect_match else None,
+    ):
         with ModelRunner(
             config, model_filename=str(model_file), overwrite_ckpt_check=False
         ) as runner:
-            runner.predict([mzml_small], result_file, evaluate=True)
+            runner.predict(files, result_file, evaluate=True)
 
-    with pytest.raises(TypeError, match=exception_string):
-        with ModelRunner(
-            config, model_filename=str(model_file), overwrite_ckpt_check=False
-        ) as runner:
-            runner.predict([mgf_small_unannotated], result_file, evaluate=True)
-
-    with pytest.raises(TypeError, match=exception_string):
-        with ModelRunner(
-            config, model_filename=str(model_file), overwrite_ckpt_check=False
-        ) as runner:
-            runner.predict(
-                [mgf_small_unannotated, mzml_small], result_file, evaluate=True
-            )
-
-    # MzTab with just metadata is written in the case of FileNotFound
-    # or TypeError early exit
+    # Metadata-only MzTab should still be written
     assert result_file.is_file()
-    result_file.unlink()
-
-    # Test mix of annotated an unannotated peak files
-    with pytest.raises(TypeError):
-        with ModelRunner(
-            config, model_filename=str(model_file), overwrite_ckpt_check=False
-        ) as runner:
-            runner.predict([mgf_small, mzml_small], result_file, evaluate=True)
-
-    assert result_file.is_file()
-    result_file.unlink()
-
-    with pytest.raises(TypeError, match=exception_string):
-        with ModelRunner(
-            config, model_filename=str(model_file), overwrite_ckpt_check=False
-        ) as runner:
-            runner.predict(
-                [mgf_small, mgf_small_unannotated], result_file, evaluate=True
-            )
-
-    assert result_file.is_file()
-    result_file.unlink()
-
-    with pytest.raises(TypeError, match=exception_string):
-        with ModelRunner(
-            config, model_filename=str(model_file), overwrite_ckpt_check=False
-        ) as runner:
-            runner.predict(
-                [mgf_small, mgf_small_unannotated, mzml_small],
-                result_file,
-                evaluate=True,
-            )
-
-    result_file.unlink()
 
 
 def test_metrics_logging(tmp_path, mgf_small, tiny_config):

